pkg/test/e2e/ccm-migration/ccm_migration_test.go

```diff
@@ -40,117 +40,156 @@ import (
-			})).NotTo(gomega.HaveOccurred())
+		ginkgo.It("migrating cluster to external CCM", func() {
+			testBody(clusterJig, openstackCluster, userClient)
 		})
 	})
 })
+
+func setupAndGetUserClient(clusterJig providers.ClusterJigInterface, cluster *kubermaticv1.Cluster, clusterClientProvider *clusterclient.Provider) ctrlruntimeclient.Client {
+	ginkgo.By("setting up cluster")
+	gomega.Expect(clusterJig.Setup()).NotTo(gomega.HaveOccurred(), "user cluster should deploy successfully")
+	clusterJig.Log().Debugw("Cluster set up", "name", clusterJig.Name())
+	gomega.Expect(clusterJig.Seed().Get(context.TODO(), types.NamespacedName{Name: clusterJig.Name()}, cluster)).NotTo(gomega.HaveOccurred())
+
+	var userClient ctrlruntimeclient.Client
+	gomega.Expect(wait.Poll(utils.UserClusterPollInterval, utils.CustomTestTimeout, func() (done bool, err error) {
+		gomega.Expect(clusterJig.Seed().Get(context.TODO(), types.NamespacedName{Name: clusterJig.Name()}, cluster)).NotTo(gomega.HaveOccurred())
+		userClient, err = clusterClientProvider.GetClient(context.TODO(), cluster)
+		if err != nil {
+			clusterJig.Log().Debug("user cluster client get failed: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})).NotTo(gomega.HaveOccurred())
+	clusterJig.Log().Debugw("User cluster client correctly created")
+
+	gomega.Expect(clusterv1alpha1.AddToScheme(userClient.Scheme())).NotTo(gomega.HaveOccurred())
+
+	gomega.Expect(wait.Poll(utils.UserClusterPollInterval, utils.CustomTestTimeout, func() (done bool, err error) {
+		if err = clusterJig.CreateMachineDeployment(userClient); err != nil {
+			clusterJig.Log().Debug("machine deployment creation failed: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})).NotTo(gomega.HaveOccurred())
+	clusterJig.Log().Debug("MachineDeployment created")
+
+	return userClient
+}
+
+func testBody(clusterJig providers.ClusterJigInterface, cluster *kubermaticv1.Cluster, userClient ctrlruntimeclient.Client) {
+	ginkgo.By("enabling externalCloudProvider feature")
+	gomega.Expect(clusterJig.Seed().Get(context.TODO(), types.NamespacedName{Name: clusterJig.Name()}, cluster)).NotTo(gomega.HaveOccurred())
+	newCluster := cluster.DeepCopy()
+	newCluster.Spec.Features = map[string]bool{
+		kubermaticv1.ClusterFeatureExternalCloudProvider: true,
+	}
+	gomega.Expect(clusterJig.Seed().Patch(context.TODO(), newCluster, ctrlruntimeclient.MergeFrom(cluster))).NotTo(gomega.HaveOccurred())
+
+	ginkgo.By("getting the patched cluster")
+	annotatedCluster := &kubermaticv1.Cluster{}
+	gomega.Expect(clusterJig.Seed().Get(context.TODO(), types.NamespacedName{Name: cluster.Name}, annotatedCluster)).NotTo(gomega.HaveOccurred())
+
+	ginkgo.By("asserting the annotations existence in the cluster")
+	_, ccmOk := annotatedCluster.Annotations[kubermaticv1.CCMMigrationNeededAnnotation]
+	_, csiOk := annotatedCluster.Annotations[kubermaticv1.CSIMigrationNeededAnnotation]
+	gomega.Expect(ccmOk && csiOk).To(gomega.BeTrue())
```

Comment:
Should we wait for those annotations in some loop (e.g. `wait.Poll`)? What if it takes a few seconds for the controller to reconcile the cluster for some reason?